lib/src/main/java/io/ably/lib/transport/ConnectionManager.java

```diff
@@ -836,8 +836,13 @@ private synchronized ConnectionStateChange setState(ITransport transport, StateI
             return null;
         }
 
-        if (stateIndication.state == ConnectionState.suspended || stateIndication.state == ConnectionState.connected) {
-            this.disconnectedRetryCount = 0;
+        if (stateIndication.state == ConnectionState.connected || stateIndication.state == ConnectionState.suspended) {
+            this.disconnectedRetryAttempt = 0;
+        }
+
+        if (stateIndication.state == ConnectionState.disconnected) {
+            states.get(ConnectionState.disconnected).timeout =
+                ReconnectionStrategy.getRetryTime(ably.options.disconnectedRetryTimeout, ++disconnectedRetryAttempt);
```

Comment:
We were mutating the state map either way, inside the `enact` method itself
https://github.com/ably/ably-java/pull/955/commits/e57f82f105dd2cef5ba946500674a862d27c6df1
As @AndyTWF mentioned, it was causing an edge case where for the first retry, it was using an old timeout value.

Reply:
I am not sure what you meant by changing `ConnectionStateChange.timeout`. 
I would like to point out, the implementation in `java` and `dotnet` is completely different for setting timers from `js` one.
`ConnectionStateChange.timeout` just denotes a timeout value for the changed state, responsible for emitting notifs to external clients, I think it shouldn't be responsible for setting internal timers.

Reply:
Implementation would have been very simple if we go by your approach.
Unfortunately, timers work differently in `ably-java` implementation : ( 
Being familiar with this code, @AndyTWF would agree to this.

Reply:
I will point our where exactly timers are set into wait state and I don't think we can suspend it in between, like we cancel timers in `js` -> https://github.com/ably/ably-java/blob/41b42f42b4ee2dabc51c86438b1783fabfd76723/lib/src/main/java/io/ably/lib/transport/ConnectionManager.java#L713

Reply:
This too happens inside a `locked` state. Where no other thread can interrupt the flow without obtaining the lock.

Reply:
ok well that's very annoying. i still think we need to refactor this because mutating `ConnectionManager.states` to communicate timeout changes is not the way to go, but i guess it's gonna take a bit of work so it can wait for now.